### PR TITLE
linear model, tested locally

### PR DESCRIPTION
**Pull request by chrisying**
Implemented linear model with REINFORCE trainer (using our loss "hack" to avoid using gradients). I've tested the all methods locally and Tensorboard seems to work for all variables except the bias variables (for some reason histogram doesn't show anything for bias).

### CODE REVIEW

File: linear_model.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+        # fc1
+        self.fc1 = self._fc_layer(self.input_state, 64, 'fc1')
+
+        # fc2
+        self.fc2 = self._fc_layer(self.fc1, 64, 'fc2')
+
+        # output (fc without relu)
+        out_weight = tf.Variable(tf.truncated_normal([64, self.output_size]), name='out_weight')
+        out_bias = tf.Variable(tf.constant(0.1, shape=[self.output_size]), name='out_bias')
```
Comment (AnanyaKumar) [inline linear_model.py:90]: Might this make the car go straight from the start (check out_bias below too)? Because the e.v. of the outputs will be positive.

File: linear_model.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        self.out = tf.matmul(self.fc2, out_weight)
+        self.out = tf.nn.bias_add(self.out, self.out_bias)
```
Comment (AnanyaKumar) [inline linear_model.py]: I'm not sure if you defined out_bias, I get an error on this line

File: linear_model.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        action: B x A   sampled action using mean from pred
+        var: B x A      variance used for sampling action (MUST BE POSTIIVE)
+        reward: B x 1   reward from environment for the performed action
+
+        inputs expected to be numpy arrays
+        """
+
+        # TODO: for actor-critic, pred should be predicted from a separate critic network instead
+
+        target = (action - pred) * reward / var
```
Comment (AnanyaKumar) [inline linear_model.py]: I think the target is incorrect if we're using this method of training (based on loss), see Facebook message.

**Review by AnanyaKumar — CHANGES_REQUESTED** [review]
Thanks for adding this! The code doesn't run yet but I left some comments :)

**Comment by AnanyaKumar** [comment]
Also, make sure to update with Edgar's updates.

File: linear_model.py (hunk 1 above)
Comment (chrisying) [reply linear_model.py:90]: It's intended that the bias is initialized with a slight positive. From my experience, this is pretty standard policy since we want the activations to be non-negative. In the grand scheme of things, this doesn't really matter.

File: linear_model.py (hunk 3 above)
Comment (chrisying) [reply linear_model.py]: Yeah, I'm missing a 2 in the denominator, but the rest of it is actually correct. If I'm interpreting things correctly, the pred, is the predicted mean (it's a B x 2 vector) which includes both the predicted mean accel_x and mean accel_y. Likewise, action also includes both accel_x and accel_y already (see the doc string).

EDIT: oh also missing a pred term, added and should match what you messaged me now.

File: linear_model.py (hunk 2 above)
Comment (chrisying) [reply linear_model.py]: woops, remnant of some debugging work

File: linear_model.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        action: B x A   sampled action using mean from pred
+        var: B x A      variance used for sampling action (MUST BE POSTIIVE)
+        reward: B x 1   reward from environment for the performed action
+
+        inputs expected to be numpy arrays
+        """
+
+        # TODO: for actor-critic, pred should be predicted from a separate critic network instead
+
+        target = (action - pred) * reward / (2.0 * var)
```
Comment (AnanyaKumar) [inline linear_model.py]: I think this should be r (action - pred)/(2.0 * var) - pred? I might have made a mistake. 
The argument is that the derivative (with respect to theta) of (pred(state) - target)^2 is 2 * (pred(state) - target) * grad(pred(state)). Train will take a step in the direction of -grad = (target - pred(state))^2 is 2 * (pred(state) - target) * grad(pred(state)). So we want 2 * (target - pred(state)) = (action - pred) * reward / var. Solving this, we get what I put above? I could be making a mistake.

**Review by AnanyaKumar — COMMENTED** [review]
Thanks for the updates, did you get a chance to test this? I get an error about train_on_batch (I think run_nn just needs to be changed so that we call train on each sample individually)?

File: linear_model.py (hunk 1 above)
Comment (AnanyaKumar) [reply linear_model.py:90]: Oh ok cool, thanks!

File: linear_model.py (hunk 4 above)
Comment (chrisying) [reply linear_model.py]: Just recreating the derivation to confirm:

We want the derivative of -(pred - target)^2 to equal (action - pred) * r / var * grad(pred). (NOTE: there's a negative because optimizers are minimizers, which means they SUBTRACT the gradient from the parameter. However, from the lecture notes it looks like the policy gradient is ADDING to the parameter (see MC policy gradient here: http://www.cs.cmu.edu/~rsalakhu/10703/Lecture_PG.pdf))

The derivative is -2(pred - target)grad(pred) = (action - pred) * r / var. Therefore target should be

target = [(action - pred) * r / (2var)] + pred

I'm not very confident on this but it seems to confirm what you are saying. I've updated the code.

Comment (ddirenzo) [reply linear_model.py]: ^ Chris is right, it should be plus for the reason he mentioned.

**Comment by AnanyaKumar** [comment]
Yeah this looks right
